openlr/openlr_match_quality/openlr_assessment_tool/traffic_mode.cpp

```diff
@@ -152,72 +218,265 @@ QVariant TrafficMode::data(const QModelIndex & index, int role) const
+size_t TrafficMode::GetPointsCount() const
 {
-  if (!index.isValid() || role != Qt::EditRole)
-    return false;
+  return m_goldenPath.size();
+}
 
-  auto const newValue = value.toString();
-  auto evaluation = openlr::ParseItemEvaluation(newValue.toStdString());
-  if (evaluation == openlr::ItemEvaluation::NotAValue)
-    return false;
+m2::PointD const & TrafficMode::GetPoint(size_t const index) const
+{
+  return m_goldenPath[index].GetCoordinate();
+}
 
-  m_decodedSample->m_decodedItems[index.row()].m_evaluation = evaluation;
-  emit dataChanged(index, index);
-  return true;
+m2::PointD const & TrafficMode::GetLastPoint() const
+{
+  CHECK(!m_goldenPath.empty(), ("Attempt to get point from an empty path."));
+  return m_goldenPath.back().GetCoordinate();
+}
+
+std::vector<m2::PointD> TrafficMode::GetCoordinates() const
+{
+  std::vector<m2::PointD> coordinates;
+  for (auto const & roadPoint : m_goldenPath)
+    coordinates.push_back(roadPoint.GetCoordinate());
+  return coordinates;
+}
+
+// TODO(mgsergio): Draw the first point when the path size is 1.
+void TrafficMode::HandlePoint(m2::PointD clickPoint, Qt::MouseButton const button)
+{
+  if (!m_buildingPath)
+    return;
+
+  auto const currentPathLength = GetPointsCount();
+  auto const lastClickedPoint = currentPathLength != 0
+      ? GetLastPoint()
+      : m2::PointD::Zero();
+
+  // Get candidates and also fix clickPoint to make it more accurate and sutable for
+  // GetOutgoingEdges.
+  auto const & candidatePoints = m_pointsDelegate->GetFeaturesPointsByPoint(clickPoint);
+  if (candidatePoints.empty())
+    return;
+
+  auto reachablePoints = GetReachablePoints(clickPoint, GetCoordinates(), *m_pointsDelegate,
+                                            0 /* lookBackIndex */);
+  auto const & clickablePoints = currentPathLength != 0
+      ? GetReachablePoints(lastClickedPoint, GetCoordinates(), *m_pointsDelegate,
+                           1 /* lookBackIndex */)
+      // TODO(mgsergio): This is not quite correct since view port can change
+      // since first call to visualize points. But it's ok in general.
+      : m_pointsDelegate->GetAllJunctionPointsInViewPort();
+
+  using ClickType = PointsControllerDelegateBase::ClickType;
+  switch(m_pointsDelegate->CheckClick(clickPoint, lastClickedPoint, clickablePoints))
+  {
+    case ClickType::Add:
+      LOG(LDEBUG, ("Adding point", clickPoint));
+      // TODO(mgsergio): Think of refactoring this with if (accumulator.empty)
+      // instead of pushing point first ad then removing last selection.
+      PushPoint(clickPoint, candidatePoints);
+
+      if (currentPathLength > 0)
+      {
+        // TODO(mgsergio): Should I remove lastClickedPoint from clickablePoints
+        // as well?
+        reachablePoints.erase(
+            remove_if(begin(reachablePoints), end(reachablePoints),
+                      [&lastClickedPoint](m2::PointD const & p) {
+                        return p.EqualDxDy(lastClickedPoint, 1e-6);
```

Comment:
I don't remember quite good what's going on here, so I'd better leave it as is until next change.